Added site sharing

jwest60 opened the pull request:
You can now share sites via the plots view page.

Underneath actions, there's a new sharing card where you can send out invitations to people with whom you would like to share your sites. 

![image](https://user-images.githubusercontent.com/32902460/87088056-2255e880-c202-11ea-92f2-6f36da8e854b.png)

After inviting someone, they will receive an e-mail informing that they've been invited and will have one week to accept the invite before it expires. Pending invitations can be viewed in the site card. When they have been accepted, they will be moved under 'Shared with'.

After they have accepted the invitation, the site owner can modify the privilege levels of the user to allow them to view / edit the site.

On the sites page, by default, you will now be able to see all of your sites as well as the sites that have been shared with you. You can filter by only shared sites or only your sites using the site type filter in the top right.

![image](https://user-images.githubusercontent.com/32902460/87088349-9ee8c700-c202-11ea-9689-153a2105046d.png)

jwest60 commented:
#253

jwest60 commented:
Removed the ability to edit viewing permissions with a user with whom a site is shared because it didn't make sense

almasaeed2010 commented on app/Http/Controllers/Traits/ListsSites.php:
```diff
@@ -51,7 +60,20 @@ public function getSites($sites)
+                        $query->withShared($request->user())->orWhere('user_id', $request->user()->id);
+                        break;
+                    case 'shared':
+                        $query->withShared($request->user());
```
withShared doesn't only include shared sites, it includes all sites. So you probably want 2 scopes, scopeShared(), scopeWithShared(). This applies to the lines above as well where you add orWhere('user_id', ...). That's not needed since withShared already adds that.

jwest60 replied:
```
public function scopeWithShared($query, User $user) {
        return $query->whereHas('userSites', function($query) use ($user) {
            $query->where('user_id', $user->id);
        });
    }
```
I figured it made the most sense to do it this way since adding the `orWhere` to the above would cause sites that weren't shared (but belonged to the user) to appear in the collection

almasaeed2010 commented on app/Http/Controllers/UserSiteController.php:
```diff
@@ -0,0 +1,91 @@
+        $users = $users->get();
+
+        $users->transform(function (User $user) use ($request) {
+            $user_site = UserSite::where('user_id', $user->id)->where('site_id', $request->site_id)->first();
```
should use with instead because this means a query for each individual user. A bit slow once you have too many users.

almasaeed2010 commented on app/Http/Controllers/UserSiteController.php:
```diff
@@ -0,0 +1,91 @@
+            });
+        }
+
+        $users = $users->get();
```
Too many users if a search term is not provided. Slow if you have too many users and the browser might not be able to handle that many results. Use paginate to limit the results.

almasaeed2010 commented on app/Http/Controllers/UserSiteController.php:
```diff
@@ -0,0 +1,91 @@
+            'search' => 'nullable|max:255',
+        ]);
+
+        $users = User::where('id', '!=', $user->id)->orderBy('name', 'asc');
```
Protect the user's privacy by selecting only the columns you need (ie, id and name). This exposes user emails when inspecting the request using Chrome dev tools.

almasaeed2010 commented on resources/views/mail/site_invitation.blade.php:
```diff
@@ -0,0 +1,13 @@
+@component('mail::message')
+Hello{{ $invitation->recipient->name ? ' ' . $invitation->recipient->name : '' }},
```
Is there a chance a recipient is not present? I'd rather throw an exception and fail if a recipient does not exist.

almasaeed2010 replied:
This can be checked in the Mail class.

jwest60 replied:
I don't think it's possible for a recipient to not be present. Tested by attempting to create a SiteInvitation with no recipient (error: no default value), attempting to update an existing site invitation by setting the recipient to null (error: cannot be null), attempting to update an existing site invitation by setting the recipient to a non-existent id (error: foreign key constraint fails), and by sending an invitation, then deleting the user it was sent to (invitation is deleted when the recipient is deleted).

almasaeed2010 commented on app/Http/Controllers/UserSiteController.php:
```diff
@@ -0,0 +1,91 @@
+        return $this->deleted();
+    }
+
+    public function users(Request $request)
```
doc block please

almasaeed2010 commented on app/Policies/UserSitePolicy.php:
```diff
@@ -0,0 +1,24 @@
+     */
+    public function delete(User $user, UserSite $user_site)
+    {
+        return $user->owns($user_site->site);
```
The user who received the invitation can also decline/delete the invitation.

jwest60 commented:
Added an interface that allows users to accept / reject incoming site sharing invitations.

![image](https://user-images.githubusercontent.com/32902460/87348249-6e669d00-c522-11ea-890c-b6415d85be35.png)

almasaeed2010 commented on app/Http/Controllers/UserSiteController.php:
```diff
@@ -0,0 +1,116 @@
+     * @return \Illuminate\Http\JsonResponse|\Illuminate\Http\Response
+     */
+    public function sharedUsers(Request $request)
+    {
```
there is no authorization logic. You should protect this route from being accessed by non owners of the site.

almasaeed2010 commented on app/Http/Controllers/UserSiteController.php:
```diff
@@ -0,0 +1,116 @@
+
+        $this->validate($request, [
+            'site_id' => 'nullable|exists:sites,id',
+            'search' => 'nullable|max:255',
```
Let's make this required and min of 2 characters (needless to say the js side will need altering)

jwest60 commented:
- Moved the invite button to the right of the search bar
![image](https://user-images.githubusercontent.com/32902460/87455419-69adf180-c5d3-11ea-89c6-988c831b0d55.png)
- Changed `Shared only` to `Owned by others` and `My sites only` to `Owned by me`
- Users will only be loaded once at least two characters have been entered into the search
- Made error messages more clear
- Email link now links to the sites page and any session info has been removed
- Added authorization to the user queries

jwest60 commented:
#276 

Moved the invitation form to its own modal

![image](https://user-images.githubusercontent.com/32902460/87584275-3e96d100-c6ab-11ea-85f5-41e1651a6999.png)

![image](https://user-images.githubusercontent.com/32902460/87584315-4d7d8380-c6ab-11ea-8d3d-3777f3d3e74c.png)

When the user accepts the invitation they will automatically receive the correct editing permissions.

almasaeed2010 commented on app/Http/Controllers/Traits/ListsSites.php:
```diff
@@ -51,7 +60,20 @@ public function getSites($sites)
+                switch ($request->site_type)
+                {
+                    case 'all':
+                        $query->withShared($request->user())->orWhere('user_id', $request->user()->id);
```
Whenever adding orWhere, always group the query in a `where(function($query) ...)`. Without grouping this could be a dangerous query. (if you need more info let me know)

```suggestion
                        $query->where(function($query) use ($request) {
                            $query->withShared($request->user());
                            $query->orWhere('user_id', $request->user()->id);
                        });
```

almasaeed2010 commented on app/SiteInvitation.php:
```diff
@@ -0,0 +1,169 @@
+        return $this->belongsTo(Site::class);
+    }
+
+    public function recipient()
```
```suggestion
    /**
     * @return \Illuminate\Database\Eloquent\Relations\BelongsTo
     */
    public function recipient()
```

almasaeed2010 commented on app/UserSite.php:
```diff
@@ -0,0 +1,33 @@
+        'site_id',
+        'editable',
+    ];
+
```
Cast the editable variable to boolean so it's easy to use and more accurate in the policies.

```suggestion
    /**
     * @var string[]
     */
    protected $casts = [
        'editable' => 'boolean',
    ];
    
```

almasaeed2010 commented on database/migrations/2020_07_07_155254_create_user_sites_table.php:
```diff
@@ -0,0 +1,45 @@
+    {
+        Schema::create('user_sites', function (Blueprint $table) {
+            $table->bigIncrements('id');
+            $table->unsignedBigInteger('user_id')->nullable();
```
Is there a need for nullable here? Is it possible that a site shared doesn't have a user?


Also, just a note that you can now combine the foreign key creation with the column creation line as follows:

```php
# We can add or remove nullable as needed
$table->foreignId('user_id')->nullable()->constrained()->onDelete('cascade');
```

This would eliminate the need for:
```php
$table->foreign('user_id')
                 ->references('id')
                 ->on('users')
                 ->onDelete('cascade');
```

almasaeed2010 commented on database/migrations/2020_07_07_155254_create_user_sites_table.php:
```diff
@@ -0,0 +1,45 @@
+        Schema::create('user_sites', function (Blueprint $table) {
+            $table->bigIncrements('id');
+            $table->unsignedBigInteger('user_id')->nullable();
+            $table->unsignedBigInteger('site_id')->nullable();
```
Same as above, do we really need this to be nullable?

almasaeed2010 commented on database/migrations/2020_07_08_154837_create_site_invitations_table.php:
```diff
@@ -0,0 +1,38 @@
+    {
+        Schema::create('site_invitations', function (Blueprint $table) {
+            $table->id();
+            $table->foreignId('user_id')->constrained()->onDelete('cascade');
```
Oh you are using the new format here 👍

almasaeed2010 commented on database/migrations/2020_07_15_162657_add_user_edit_column_to_site_invitations_table.php:
```diff
@@ -0,0 +1,32 @@
+    public function up()
+    {
+        Schema::table('site_invitations', function (Blueprint $table) {
+            $table->boolean('user_edit')->default(false);
```
For future reference, boolean values in a table would conventionally be named as a question. i.e. is_editable, or is_closed etc...

You don't have to change this. I am just pointing out the convention here.

almasaeed2010 commented on resources/js/components/Dropdown.vue:
```diff
@@ -35,9 +35,16 @@
+            <p v-else-if="options.length === 0 && emptyMessage.length"
                class="text-center text-muted p-2 mb-0">{{ emptyMessage }}</p>
-            <p v-if="endMessage.length" class="text-center text-muted p-2 mb-0">
+            <p v-else="endMessage.length" class="text-center text-muted p-2 mb-0">
```
Did you mean `v-else-if` or should you remove the param? I don't think v-else takes a parameter.

almasaeed2010 commented on resources/js/components/Dropdown.vue:
```diff
@@ -134,6 +144,7 @@
       },
 
       down() {
+        console.log(this.total)
```
```suggestion
```

almasaeed2010 commented on resources/js/components/SiteSharingCard.vue:
```diff
@@ -0,0 +1,290 @@
+      async toggleEdit(user) {
+        try {
+          const {data} = await axios.put(`/web/user-sites/site/${this.site.id}/user/${user.id}/toggle-edit`)
+          user.can_edit = data.editable
```
What are we trying to do here? if you want to update the user instance without having to reload, then you have to alter the array itself. This is because Vue only watches objects in a shallow manner. ie, it can't detect changes to a property of an object that is part of an array. Therefore, the UI won't always get updated using this method.

```suggestion
          this.sharedWith = this.sharedWith.map(u => u.id === user.id ? {...u, can_edit: data.editable} : u)
```

almasaeed2010 commented on resources/js/components/SiteSharingCard.vue:
```diff
@@ -0,0 +1,290 @@
+              text: 'You are not authorized to complete this action.',
+              type: 'error',
+            })
+          } else if (e.response && e.response.status === 422) {
+            const errors = new Errors(e.response.data)
+            this.$alert(errors.toArray().join(' '))
+          } else {
```
Since it's not possible to receive a 422 when there is no validation, this bit of error detection is not actually needed.

```suggestion
          } else {
```

almasaeed2010 commented on routes/web.php:
```diff
@@ -64,6 +64,9 @@
 
 Route::get('/invitations/{invitation}/accept', 'InvitationController@accept');
 
+Route::get('/site-invitations/{invitation}/accept', 'SiteInvitationController@accept');
+Route::get('/site-invitations/{invitation}/reject', 'SiteInvitationController@reject');
+
```
We no longer need this.

```suggestion
```

jwest60 commented:
I applied the majority of the code review changes and removed any migrations that added / dropped columns in favor of just modifying the create table migration since we haven't deployed yet. I also re-worked the accept / reject functions to be general validation for the site invitation (i.e. is this site invitation expired? has it been processed? is the person accepting it the correct recipient?) before accepting / rejecting the invitation.

jwest60 commented:
Moved the accept / reject routes to the verified middleware and adjusted any GET requests to those routes accordingly.

almasaeed2010 commented on app/Http/Controllers/SiteController.php:
```diff
@@ -37,9 +37,12 @@ public function index(Request $request, ?User $user = null)
         ]);
 
-        $sites = $this->getSites($user->sites())->paginate($request->limit ?? 20);
+        $sites = null;
```
This is immediately overridden below.

```suggestion
```

almasaeed2010 commented on app/Http/Controllers/SiteInvitationController.php:
```diff
@@ -0,0 +1,190 @@
+     * @param \Illuminate\Http\Request $request
+     * @return \Illuminate\Contracts\Foundation\Application|\Illuminate\Contracts\View\Factory|\Illuminate\Http\RedirectResponse|\Illuminate\View\View|void
+     */
+    public function accept(SiteInvitation $invitation, Request $request)
```
@jwest60 can we remove these or are they needed?
